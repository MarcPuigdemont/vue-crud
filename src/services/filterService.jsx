const contains = (string, substring) => string.toLowerCase().indexOf(substring.toLowerCase()) > -1;

const filterByStringAndServices = (airlines, string, serviceList = []) => {
  const matchingAirlines = string.trim() === '' ? airlines : airlines.filter(a => contains(a.iata, string) || contains(a.name, string));

  const numOfServices = serviceList.length;
  return matchingAirlines.filter(airline => serviceList.filter(s => airline.services[s]).length === numOfServices);
};

// eslint-disable-next-line import/prefer-default-export
export { filterByStringAndServices };
