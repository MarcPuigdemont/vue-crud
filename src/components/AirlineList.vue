<template>
  <b-list-group class="w-100">
    <b-row no-gutters class="pl-3 font-weight-bold">
      <b-col>IATA</b-col>
      <b-col>Name</b-col>
      <b-col>Primary Color</b-col>
      <b-col>Secondary Color</b-col>
      <b-col>Available Services</b-col>
    </b-row>
    <AirlineListItem v-for="(airline, index) in filteredAirlines" :key="index" :airline="airline" @edit="edit" />
  </b-list-group>
</template>
<script>
import { BListGroup, BCol, BRow } from 'bootstrap-vue';
import AirlineListItem from './AirlineListItem.vue';
import { filterByStringAndServices } from '@/services/filterService.jsx';

export default {
  name: 'AirlineList',
  components: {
    AirlineListItem,
    BListGroup,
    BCol,
    BRow,
  },
  props: {
    airlines: Array,
    filter: String,
    serviceFilter: Array,
  },
  computed: {
    filteredAirlines() {
      return filterByStringAndServices(this.airlines, this.filter, this.serviceFilter);
    },
  },
  methods: {
    edit(airline) {
      this.$emit('edit', airline);
    },
  },
};
</script>
