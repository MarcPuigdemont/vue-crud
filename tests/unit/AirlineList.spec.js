import { mount, shallowMount } from '@vue/test-utils';
import AirlineList from '@/components/AirlineList.vue';
import * as filterService from '@/services/filterService.jsx';

const propsData = {
  airlines: [
    {
      iata: 'BX',
      name: 'Airline name',
      primary_color: '#ffffff',
      secondary_color: '#000000',
      services: { bags: true, checkin: false, seats: true },
    },
    {
      iata: 'AX',
      name: 'Airline name 2',
      primary_color: '#ffaaff',
      secondary_color: '#00aa00',
      services: { bags: false, checkin: true, seats: true },
    },
  ],
  filter: '',
  serviceFilter: [],
};

describe('AirlineList', () => {
  test('should match the snapshot', () => {
    const wrapper = mount(AirlineList, {
      propsData,
      stubs: {
        AirlineListItem: '<div class="airline-grid-item"></div>',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  test('should emit the same airline as it gets from edit signal', () => {
    filterService.filterByStringAndServices = jest.fn();

    const wrapper = shallowMount(AirlineList, {
      propsData,
    });

    wrapper.vm.edit(propsData.airlines[0]);
    expect(wrapper.emitted('edit')[0][0]).toEqual(propsData.airlines[0]);
    expect(filterService.filterByStringAndServices).toBeCalledWith(propsData.airlines, propsData.filter, propsData.serviceFilter);
  });
});
