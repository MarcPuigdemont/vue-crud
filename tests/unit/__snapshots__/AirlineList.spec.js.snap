// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AirlineList should match the snapshot 1`] = `
<div
  class="list-group w-100"
>
  <div
    class="row pl-3 font-weight-bold no-gutters"
  >
    <div
      class="col"
    >
      IATA
    </div>
     
    <div
      class="col"
    >
      Name
    </div>
     
    <div
      class="col"
    >
      Primary Color
    </div>
     
    <div
      class="col"
    >
      Secondary Color
    </div>
     
    <div
      class="col"
    >
      Available Services
    </div>
  </div>
   
  <div
    class="airline-grid-item"
  />
  <div
    class="airline-grid-item"
  />
</div>
`;
